src/utils.cpp

```diff
@@ -360,4 +361,353 @@ int mega_snprintf(char *s, size_t n, const char *format, ...)
+        offset += 2;
+
+        // if there's not enough data for value...
+        if (offset + valuelen > datalen)
+        {
+            delete tlv;
+            return NULL;
+        }
+
+        // get the Value
+        value.resize(valuelen);
+        value.assign((char*)&(data->data()[offset]), valuelen);  // value may include NULL characters, read as a buffer
+        offset += valuelen;
+
+        // add it to the map
+        tlv->set(type, value);
+    }
+
+    return tlv;
+}
+
+
+TLVstore * TLVstore::containerToTLVrecords(const string *data, SymmCipher *key)
+{
+    if (data->empty())
+    {
+        return NULL;
+    }
+
+    unsigned offset = 0;
+    encryptionsetting_t encSetting = (encryptionsetting_t) data->at(offset);
+    offset++;
+
+    unsigned ivlen = TLVstore::getIvlen(encSetting);
+    unsigned taglen = TLVstore::getTaglen(encSetting);
+    encryptionmode_t encMode = TLVstore::getMode(encSetting);
+
+    if (!ivlen || !taglen || encMode == AES_MODE_UNKNOWN)
+    {
+        return NULL;
+    }
+
+    byte *iv = new byte[ivlen];
+    memcpy(iv, &(data->data()[offset]), ivlen);
+    offset += ivlen;
+
+    unsigned cipherTextLen = data->length() - offset;
+    string cipherText = data->substr(offset, cipherTextLen);
+
+    unsigned clearTextLen = cipherTextLen - taglen;
+    string clearText;
+
+    if (encMode == AES_MODE_CCM)   // CCM or GCM_BROKEN (same than CCM)
+    {
+        key->ccm_decrypt(&cipherText, iv, ivlen, taglen, &clearText);
+    }
+    else if (encMode == AES_MODE_GCM)  // GCM
+    {
+        key->gcm_decrypt(&cipherText, iv, ivlen, taglen, &clearText);
+    }
+
+    delete [] iv;
+
+    TLVstore *tlv = TLVstore::containerToTLVrecords(&clearText);
+    if (!tlv) // 'data' might be affected by the legacy bug: strings encoded in UTF-8 instead of Unicode
+    {
+        // retry TLV decoding after conversion from 'UTF-8 chars' to 'Unicode chars'
+        LOG_warn << "Retrying TLV records decoding with UTF-8 patch";
+
+        string clearTextUnicode;
+        if (!Utils::utf8toUnicode((const byte*)clearText.data(), clearTextLen, &clearTextUnicode))
+        {
+            LOG_err << "Invalid UTF-8 encoding";
+        }
+        else
+        {
+            tlv = TLVstore::containerToTLVrecords(&clearTextUnicode);
+        }
+    }
+
+    return tlv;
+}
+
+TLVstore::~TLVstore()
+{
+}
+
+bool Utils::utf8toUnicode(const uint8_t *src, unsigned srclen, string *result)
+{
+    uint8_t utf8cp1;
+    uint8_t utf8cp2;
+    int32_t unicodecp;
+
+    if (!srclen)
+    {
+        result->clear();
+        return true;
+    }
+
+    byte *res = new byte[srclen];
+    unsigned rescount = 0;
+
+    unsigned i = 0;
+    while (i < srclen)
+    {
+        utf8cp1 = src[i++];
+
+        if (utf8cp1 < 0x80)
+        {
+            res[rescount++] = utf8cp1;
+        }
+        else
+        {
+            if (i < srclen)
+            {
+                utf8cp2 = src[i++];
```

Comment:
It would be good to check if `utf8cp1` is `0xC2` or `0xC3` and `utf8cp2` is lower than `0xC0` and greater or equal to `0x80` (valid two-bytes UTF-8 character)